### PR TITLE
[CI Environment] Enabled CI environment to handle DeploymentTest (bicep) files for new self-contained dependency approach

### PR DESCRIPTION
**Pull request by MrMCake**
# Description

Implementing 'Self-contained' module dependency deployment for variant 5 #329

Tasks
- [x] Adjust GitHub workflows
- [x] Adjust Azure DevOps pipelines
- [x] Update Set-ModuleReadMe function to be able to convert bicep to JSON parameter
- [x] Test GitHub (with and without new type of dependency file)
- [x] Test ADO (with and without new type of dependency file)
- [x] Test pure ARM execution & utilities
- [x] Test Azure DevOps dependency pipeline
- [x] Test GitHub dependency pipeline
- [x] Update docs

> - Note: #1612 should be merged first
> - Note: #1632 should be merged first
> - Note: #1672 should be merged first
> - Note: #1678 should be merged first
> - Note: #1694 should be merged first
> - Note: #1516 should be merged first

## Open questions
- Should Set-ModuleReadMe script support ARM-JSON-Module-Test files and extract 'parameters'? 

<details>
<summary>GitHub</summary>

image

</details>

<details>

<summary>Azure DevOps</summary>

image

</details>

| | GitHub Workflow | Azure DevOps Pipeline | New test files |
| - | - | - | - |
| | [Sql: Servers](https://github.com/Azure/ResourceModules/actions/workflows/ms.sql.servers.yml) | [Build Status](https://dev.azure.com/carml/CARML/_build/latest?definitionId=90&branchName=users%2Falsehr%2F1583_dependencyVar5) | ✅  |
| | [AnalysisServices: Servers](https://github.com/Azure/ResourceModules/actions/workflows/ms.analysisservices.servers.yml) | [Build Status](https://dev.azure.com/carml/CARML/_build/latest?definitionId=2&branchName=users%2Falsehr%2F1583_dependencyVar5) | ✅  |
| | [Network: VPN Sites](https://github.com/Azure/ResourceModules/actions/workflows/ms.network.vpnsites.yml) | [Build Status](https://dev.azure.com/carml/CARML/_build/latest?definitionId=79&branchName=users%2Falsehr%2F1583_dependencyVar5) | ✅  |
| | [Compute: VirtualMachines](https://github.com/Azure/ResourceModules/actions/workflows/ms.compute.virtualmachines.yml) | [Build Status](https://dev.azure.com/carml/CARML/_build/latest?definitionId=21&branchName=users%2Falsehr%2FdepBack) | ✅ |
| | [KeyVault: Vaults](https://github.com/Azure/ResourceModules/actions/workflows/ms.keyvault.vaults.yml) | [Build Status](https://dev.azure.com/carml/CARML/_build/latest?definitionId=46&branchName=users%2Falsehr%2F1583_dependencyVar5) | :paperclip: (no) |
| | [.Platform: Dependencies](https://github.com/Azure/ResourceModules/actions/workflows/platform.dependencies.yml) (unrelated) | [Build Status](https://dev.azure.com/carml/CARML/_build/latest?definitionId=105&branchName=users%2Falsehr%2F1583_dependencyVar5) |
| ARM Test | [AnalysisServices: Servers](https://github.com/Azure/ResourceModules/actions/workflows/ms.analysisservices.servers.yml) | [Build Status](https://dev.azure.com/carml/CARML/_build/latest?definitionId=2&branchName=users%2Falsehr%2F1583_armTest) | ✅ |

# Type of Change

Please delete options that are not relevant.

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Update to documentation

### CODE REVIEW

**Review by MariusStorhaug — CHANGES_REQUESTED** [review]
Some comments and suggestions.

File: docs/wiki/The CI environment - Pipeline design.md [hunk 1]
```diff
@@ -118,6 +118,8 @@ In addition to module pipelines, the repository includes several platform pipeli
 
 ## Dependencies pipeline
 
+> NOTE: The dependencies deployed as part of this pipeline will be moved to the individual modules that depend on them once issue [1583](https://github.com/Azure/ResourceModules/issues/1583) is resolved. You can find further information about this effort [here](./The%20library%20-%20Module%20design#module-test-files).
```
[the remark this answers is not in the document]
Comment (MrMCake) [reply docs/wiki/The CI environment - Pipeline design.md:121]: Added a reference to issue #1787 ([ref](https://github.com/Azure/ResourceModules/issues/1791#issuecomment-1229471784))

File: docs/wiki/The library - Module design.md [hunk 2]
```diff
@@ -524,13 +524,110 @@ Note the following recommendations:
...
-- A module should have as many parameter files as it needs to evaluate all parts of the module's functionality.
-- Sensitive data should not be stored inside the parameter file but rather be injected by the use of tokens, as described in the [Token replacement](./The%20CI%20environment%20-%20Token%20replacement) section, or via a [Key Vault reference](https://docs.microsoft.com/en-us/azure/azure-resource-manager/templates/key-vault-parameter?tabs=azure-cli#reference-secrets-with-static-id).
+# Module test files
+
+Module test files in CARML are implemented in
+- a classic way leveraging the common `deploymentParameters.json` schema for ARM deployments or
+- using comprehensive `.bicep` test files that not only test the module's template in a certain scenario, but also deploy any required dependency for it. All classic test files will be migrated to this module following the issue [1583](https://github.com/Azure/ResourceModules/issues/1583).
+
+As parameters are usually specific to their corresponding template, we only have a few general recommendations:
```
[the remark this answers is not in the document]
Comment (MrMCake) [reply docs/wiki/The library - Module design.md:533]: Added a reference to issue #1787 ([ref](https://github.com/Azure/ResourceModules/issues/1791#issuecomment-1229471784))

File: docs/wiki/The CI environment - Deployment validation.md [hunk 3]
```diff
@@ -21,19 +21,21 @@ The deployment validation phase can be divided into three steps, running in sequ
...
 
-The parameter files used in this stage should ideally cover as many configurations as possible to validate the template flexibility, i.e., to verify that the module can cover multiple scenarios in which the given Azure resource may be used. Using the example of the CosmosDB module, we may want to have one parameter file for the minimum amount of required parameters, one parameter file for each CosmosDB type to test individual configurations, and at least one parameter file testing the supported extension resources such as RBAC & diagnostic settings.
+> NOTE: Once the issue [1583](https://github.com/Azure/ResourceModules/issues/1583) is resolved, the deployment of these dependencies will be moved into the module test files. You can find additional information about this effort [here](./The%20library%20-%20Module%20design#module-test-files).
```
[the remark this answers is not in the document]
Comment (MrMCake) [reply docs/wiki/The CI environment - Deployment validation.md:36]: Added a reference to issue #1787 ([ref](https://github.com/Azure/ResourceModules/issues/1791#issuecomment-1229471784))

File: docs/wiki/The library - Module design.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (MrMCake) [reply docs/wiki/The library - Module design.md:533]: Gotcha, I wasn't aware 'we' (as a means to express 'those responsible for the platform') is a bad practice. I'll see to find a different wording

[the remark this answers is not in the document]
Comment (MariusStorhaug) [reply docs/wiki/The library - Module design.md:533]: @matebarabas Any thoughts on this one?

[the remark this answers is not in the document]
Comment (MrMCake) [reply docs/wiki/The library - Module design.md:533]: Just noticed this wasn't written by me :D

[the remark this answers is not in the document]
Comment (MrMCake) [reply docs/wiki/The library - Module design.md:533]: Done ;)

File: docs/wiki/The library - Module design.md [hunk 4]
```diff
@@ -440,11 +454,12 @@ Within a bicep file, use the following conventions:
...
-  - When a module requires to deploy a resource whose resource type is outside of the main module's provider namespace, the module of this additional resource is referenced locally. For example, when extending the Key Vault module with Private Endpoints, instead of including in the Key Vault module an ad hoc implementation of a Private Endpoint, the Key Vault directly references the Private Endpoint module (i.e., `module privateEndpoint 'https://github.com/Azure/ResourceModules/blob/main/Microsoft.Network/privateEndpoints/deploy.bicep'`). Major benefits of this implementation are less code duplication, more consistency throughout the module library and allowing the consumer to leverage the full interface provided by the referenced module.
+- Module symbolic names are in camel*Snake_Case, following the schema `<mainResourceType>*<referencedResourceType>`e.g.,`storageAccount_fileServices`, `virtualMachine_nic`, `resourceGroup_roleAssignments`.
```
Comment (MariusStorhaug) [inline docs/wiki/The library - Module design.md:460]: What is this change about? Should be '_' right?

**Review by MariusStorhaug — CHANGES_REQUESTED** [review]

File: docs/wiki/The library - Module design.md (hunk 4 above)
Comment (MariusStorhaug) [reply docs/wiki/The library - Module design.md:460]: camel*Snake_Case -> camel_Snake_Case? Also some spaces missing?
>>`<mainResourceType>*<referencedResourceType>`e.g.,<< revert this? or am I missing something?

Comment (MrMCake) [reply docs/wiki/The library - Module design.md:460]: Very curious indeed. Undoing

Comment (MrMCake) [reply docs/wiki/The library - Module design.md]: Turns out - it was a change automatically done by an extension

**Pull request merged**